### PR TITLE
CIRC-10276 - Cleanup base64 encoding

### PR DESCRIPTION
**Pull request by RobBoeckermann**
I no longer think there's an issue with the encoding or decoding. 

Using the existing integer arithmetic, `mtev_b64_encode_len` will return the *exact* length of an encoded string created from a given input length (including any padding). `mtev_b64_max_decode_len` can return up to 2 bytes extra because it does not know if the input length includes any padding.

I think the bug I was seeing in reconnoiter was caused by calling `mtev_b64_max_decode_len` directly instead of using a reconnoiter wrapper to add 3 bytes for `b""` (that still needs to be verified).

None of the changes in this PR fix any known issues, but it helps keep our code DRY and cleans things up a bit.

* call `mtev_b64_encode_len` and `mtev_b64_max_decode_len` functions instead of hard coding values or repeatedly defining logic.
* improve base64 testing.
* document additional dependency.
* add testing documentation.

### CODE REVIEW

**Review by esproul — COMMENTED** [review]
I'm probably going to end up closing #902 as a result of this, but could you please incorporate the additional changelog items I had there?

**Review by vasuraman11 — APPROVED** [review]
LGTM

**Comment by esproul** [comment]
> I'm probably going to end up closing #902 as a result of this, but could you please incorporate the additional changelog items I had there?

On second thought, nevermind. I can rebase 902 when we're ready to release.

**Comment by RobBoeckermann** [comment]
DO NOT MERGE

I need to keep looking into this. `test/lua-tests/b64_spec.lua` needs to have a test for this.
We are currently testing `A == decode(encode(A))`, but the bug that prompted this change was found when doing `A == encode(decode(A))`.
Adding a test for that fails master, as expected, but I found this comment in `mtev_b64_decode`:
```
  /* Needed here is 2 bytes shy of what could be needed...
   * decoding can be "short" up to 2 bytes. */
```
This means that it was already known that the decode length can be incorrect. The presence of a comment instead of a fix worries me because I fear that we may have worked around this bug in ways that make it dangerous to change.

**Review by esproul — CHANGES_REQUESTED** [review]

File: ChangeLog.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 # 2
 
+## 2.4.5
```
[the remark this answers is not in the document]
Comment (RobBoeckermann) [reply ChangeLog.md:5]: that doesn't follow the pattern of the rest of the doc. you want me to replace "2.4.5" with "2.4"?

[the remark this answers is not in the document]
Comment (esproul) [reply ChangeLog.md:5]: The formatting is a bit messed up right now. The top level should be the major version, then the next is major.minor, and third level should be patch. We add a new patch heading when we make a release, so unreleased changes in master should sit between the previous patch and the current major.minor.  Does that make sense?

[the remark this answers is not in the document]
Comment (RobBoeckermann) [reply ChangeLog.md:5]: Got it 👍

File: src/mtev_websocket_client.c [hunk 2]
```diff
@@ -223,11 +223,10 @@ recv_resheader(eventer_t e, char *buf, int len, int *mask) {
  * few different init phases that will have been called before us */
 static void
 mtev_websocket_client_create_key(char *dest) {
-  /* base64 encoded length is 4*ceil(n/3) so a pre-encoding length of 18 gives
-   * us an encoded length of 24 */
-  unsigned char buf[18];
+  const size_t decoded_length = 18;
```
Comment (RobBoeckermann) [inline src/mtev_websocket_client.c:226]: use `dest` length instead of hard coded value?

File: src/modules/lua_mtev.c [hunk 3]
```diff
@@ -3252,13 +3252,12 @@ nl_hmac_sha1_encode(lua_State *L) {
...
 
   if(lua_gettop(L) != 2) luaL_error(L, "bad call to mtev.hmac_sha1_encode");
-  encoded_len = 28; /* the length of the base64 encoded HMAC-SHA1 result will always be 28 */
 
   message = (const unsigned char *)lua_tolstring(L, 1, &messagelen);
   key = (const unsigned char *)lua_tolstring(L, 2, &keylen);
 
   HMAC(EVP_sha1(), key, keylen, message, messagelen, result, &md_len);
-  encoded_len = mtev_b64_encode(result, md_len, encoded, encoded_len);
+  encoded_len = mtev_b64_encode(result, md_len, encoded, mtev_b64_encode_len(md_len));
```
Comment (pjulien) [inline src/modules/lua_mtev.c:3260]: why wouldn't this be encoded_len?  It's the result of ``mtev_b64_encode_len``

**Review by pjulien — CHANGES_REQUESTED** [review]

File: src/modules/lua_mtev.c (hunk 3 above)
Comment (RobBoeckermann) [reply src/modules/lua_mtev.c:3260]: `encoded_len` is uninitialized here. It used to be hard coded to 28, but I removed that in favor of calling the function.

**Review by pjulien — APPROVED** [review]

**Review by esproul — APPROVED** [review]

**Pull request merged**